### PR TITLE
Engine abstraction - rewrite Writers class to use ResultProcessorBuilder provided by platform

### PR DESCRIPTION
**Pull request by pihme**
## Description

This is the last PR in the "shape legacy code into new interfaces" series.

Previously, the new interfaces `ProcessingResultBuilder` and `ProcessingResult` were introduced. This is how the engine is supposed to return the processing result to the stream processor. Also previously, these classes were passed between stream processor and engine on interface level, and the division of labor between stream processor and engine was established.

What was not achieved yet, is that the engine uses these new classes. Instead, the engine used existing legacy code to circumvent the new interfaces. 

This PR transforms the engine to use the new classes provided by the stream processor:
- Central point of transformation is the `Writers` class which at the outset depended on stream writers, state writers, response writers and so on
- In the end, the `Writers` class only depends on `ProcessingResultBuilder` which will be swapped out for each record processed, or each error to be handled
- To make this transition possible we a) need to refine the `Typed(Response/Rejection/Command)Writer` classes and remove methods like `reset, flush` which will no longer be offered by the new interfaces and b) need to implement the refined interfaces with an implementation that forwards calls to `ProcessingResultBuilder`

## Review hints
- @npepinpe 95% of changes is in the engine, @remcowesterhoud will look into those
- These changes are done under "better done than perfect" regime. Some naming might be not exactly right. Some clean up might be done after the fact. The focus is to establish much of the engine abstraction concept, so that both teams afterwards can work without interfering with each other.

## Related issues

related to #9725

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/Engine.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/Engine.java:95]: ❓ Why do we need this? I see it sets the mutex, but I don't understand how this mutex is working here. Does this try with resources somehow only lets 1 thread access the resource at a time?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/LegacyTypedStreamWriterProxy.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/LegacyTypedStreamWriterProxy.java:17]: ❓ Maybe I am misunderstanding the Legacy stuff. My thoughts were that these were going to be removed entirely. They are still here, do we need to keep these or will they be removed later in the abstraction topic?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/EventApplyingStateWriter.java
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/EventApplyingStateWriter.java:18]: 🔧 I can't put this comment at a different line as that code is unchanged, so I'll put it here instead. I believe this class can be removed entirely. It's only been used in tests currently.

**Review by remcowesterhoud — APPROVED** [review]
Thanks for the small commits, that made it a lot better to review this 🙇 

I'll be honest I didn't understand everything that's been changed, probably because I'm not that deep into the abstraction topic. I have added some questions and some minor suggestions.

Overall I didn't see any strange stuff and I like the look of the abstraction so far, so great job 🚀

File: engine/src/main/java/io/camunda/zeebe/engine/Engine.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/Engine.java:95]: We don't need this. And it doesn't address multthreading. The only thing the try-with-resources does is to close the scope even if an exception is thrown in the block

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/EventApplyingStateWriter.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/EventApplyingStateWriter.java:18]: Good catch!

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/LegacyTypedStreamWriterProxy.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/LegacyTypedStreamWriterProxy.java:17]: I could mark them as deprecated, but CodeQL took the fun out of it. It will report a use of a deprecated method as an issue and then you manually have to confirm that it was intended.

Eventually, yes, they will be removed. But there are more steps to take in the engine abstraction topic. To describe it in a metaphor, we have built a new bridge for traffic go across, but the old ways are still used in some places. 

For example `LegacyTypedCommandWriter` is still used in `JobTimeoutTrigger`, which implements a background task to deactivate timed out tasks. It is classes like this, which are adjacent to the engine, but not part of the engine in terms of record processing which still use the old code.

Continuation of the engine abstraction will be to refactor all these places and move them also over to the next code. But this might take a while. Consider for example #8002 which has been open for some time now.

**Review by npepinpe — APPROVED** [review]
:+1: It's great to see everything coming together/into shape!

There was only one issue, which is very minor, about a potentially unused field. But nothing that requires a second review I think.

- :thought_balloon: As I wrote, not sure about the whole mutex and if it does anything useful, but happy to see how it helps us in practice.
- :thought_balloon: It's a bit out of scope, but now that we have an expandable buffer in the result builder, could a large synchronous multi-instance collection cause us to run out of memory by writing tons of event? I guess it's not really all that different from before as we were anyway staging them in memory, right? So just for my peace of mind, the situation is more or less the same as before, correct?

File: engine/src/main/java/io/camunda/zeebe/engine/Engine.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/Engine.java:48]: The mutex is mainly to ensure that callers get the latest request-scoped builder. You are right that it doesn't prevent the result from being cached.

If you want to make sure the builder is not used out of scope, the platform could pass a proxy to the engine and invalidate the proxy after it goes out of scope. Then, even if the engine were to cache the result, you could detect it.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/ResultBuilderBackedTypedResponseWriter.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/ResultBuilderBackedTypedResponseWriter.java:35]: Copy and paste error

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/Writers.java
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/Writers.java:15]: Yes it could. Personally, not a fan of using records for something which isn't a DTO

**Comment by pihme** [comment]
> * It's a bit out of scope, but now that we have an expandable buffer in the result builder, could a large synchronous multi-instance collection cause us to run out of memory by writing tons of event? I guess it's not really all that different from before as we were anyway staging them in memory, right? So just for my peace of mind, the situation is more or less the same as before, correct?

The expandable buffer is not in use yet. We are still using the direct access. So it is exactly as before

**Comment by pihme** [comment]
bors merge

**Pull request merged**
